Capabilities 3

zippy opened the pull request:
- [x] cleanup of code paths that end up yielding capability checks
- [x] create mock CallSignature of call.
- [x] checking token provenence works (i.e. checking the signature) for Assigned capabilities
- [x] checking token value works for Transferrable tokens
- [x] checking the token value for agent works
- [x] generate public capability at genesis time for "public" trait functions
- [x] added PUBLIC_TOKEN global for access to that token for use in hdk::call 
- [x] CapbilityCall no longer optional and renamed to CapabilityRequest.
- [x] internal CapabilityRequest generated for callbacks.
- [x] passing in the correct token actually works from the container
- [x] move capability functions to correct places
- [x] condense all public functions in all zomes into one public capability token during init

The actual capabilities are hooked up in a number of ways now in that any functions that you want to be made public should be listed in the trait with the special reserved name "hc_public".  Such traits get added to a public access capability grant that's recorded in the chain at chain initialization, and which the conductor can get access to from the initialization result which it can use to pass in for the capability call.

Not in this PR and followup for Capabilities-4: 
- [ ] a `grant` zome api function for app developers to be able to make capabilities grants and `verify_grant` function for verifying capability requests.
- [ ] adding of grants at bridge_genesis time for making it easy for the conductor to connect DNA instances
- [ ] merging of the Agent ID entry and the AgentCapability.
- [ ] land how cap request will be made in `conductor_api/src/interface.rs`

sphinxc0re commented on nodejs_container/README.md:
```diff
@@ -46,8 +46,8 @@ const scenario = new Scenario([instanceAlice, instanceBob])
-    const result1 = alice.call('zome', 'capability', 'function', {params: 'go here'})
-    const result2 = bob.call('zome', 'capability', 'function', {params: 'go here'})
+    const result1 = alice.call('zome', 'function', {params: 'go here'})
+    const result2 = bob.call('zome', 'function', {params: 'go here'})
```
Don't we still need the trait here?

zippy replied:
nope, no trait ever needed for calls, only the capability token which will be in the params.

zippy commented on nodejs_container/native/src/container.rs:
```diff
@@ -0,0 +1,142 @@
+use holochain_container_api::{
```
yep, removed

zippy commented on core_types/src/dna/capabilities.rs:
```diff
@@ -49,21 +43,39 @@ impl ReservedCapabilityNames {
+}
 
 #[derive(Serialize, Deserialize, Clone, Debug, Eq, PartialEq, Hash)]
 pub struct CapabilityCall {
```
I agree.  They are just here because this is where capabilities was...

zippy replied:
I've moved CapabilityCall (and renamed CapabilityRequest) into ribosome::capability.  I've just removed CallSignature all together, it was, I think, an unnecessary wrap of Signature.  So all that's left in the old ribosome::dna::capabilites is trait stuff so I renamed that to `traits.rs`

zippy commented on core/src/nucleus/ribosome/runtime.rs:
```diff
@@ -33,7 +28,47 @@ pub struct Runtime {
+    pub fn fn_name(&self) -> String {
+        match self {
+            WasmCallData::ZomeCall(data) => data.zome_call.fn_name.clone(),
+            WasmCallData::NullCall => "__hdk_get_json_definition".to_string(),
```
Good idea.  Makes sense.  Maybe "DirectCall"

zippy commented on core/src/nucleus/ribosome/runtime.rs:
```diff
@@ -33,7 +28,47 @@ pub struct Runtime {
+#[derive(Clone)]
+pub enum WasmCallData {
+    ZomeCall(ZomeCallData),
+    NullCall,
```
fixed

zippy commented on core/src/nucleus/ribosome/fn_call.rs:
```diff
@@ -0,0 +1,1232 @@
+    call_sig.signature() == mock_signature
+}
+
+pub fn make_cap_call<J: Into<JsonString>>(
```
added some comments and renamed to make clearer.

zippy commented on core/src/nucleus/ribosome/fn_call.rs:
```diff
@@ -0,0 +1,1232 @@
+
+    let cap_type = grant.cap_type();
+    if cap_type == CapabilityType::Public {
+        return true;
```
you are right.  I removed this

zippy commented:
I renamed dna::capabilities to dna::traits, added a ribosome::capabilities, and renamed CapabilityCall to CapabilityRequest which I moved into the capabilities file.  The other capability sutff is about the fn_call use of capabilities, so I left it in `fn_call.rs` thought some parts should be refactored out..

zippy commented on core/src/nucleus/actions/initialize.rs:
```diff
@@ -96,13 +163,22 @@ pub async fn initialize_application(
         });
 
+    // otherwise return the Initialization struct
+    let initialization_result = match maybe_error {
```
:+1:

zippy commented on core/src/nucleus/actions/call_zome_function.rs:
```diff
@@ -63,44 +70,20 @@ pub async fn call_zome_function(
-        if !public && !check_capability(context.clone(), &zome_call.clone()) {
-            return Err(HolochainError::CapabilityCheckFailed);
-        }
+    // the state
```
done

zippy commented on conductor_api/src/interface.rs:
```diff
@@ -164,13 +167,28 @@ impl ConductorApiBuilder {
-                                    Some(CapabilityCall::new(Address::from("fake_token"), None)),
+
+                            let cap_call = {
+                                // TODO: get the token from the paramters on only if not there
```
cleaned up.

zippy commented:
Fixed the initialization issue.  For now everywhere we use the public token we just look it up.

lucksus commented on core/src/action.rs:
```diff
@@ -172,7 +173,7 @@ pub enum Action {
     InitApplication(Dna),
     /// return the result of an InitApplication action
     /// the result is Some arbitrary string
```
still

willemolding commented:
@zippy I was able to reproduce the problem in CI locally. In my local windows environment running `cargo test -p holochain_core --release` results in:

```

Finished release [optimized] target(s) in 0.45s                                                                                              Running target\release\deps\holochain_core-f48f6abc05c74485.exe                                                                                                                                                                                                                    running 99 tests                                                                                                                            test action::tests::action_wrapper_action ... ok                                                                                            test action::tests::action_wrapper_hash ... ok                                                                                              test action::tests::new_action ... ok                                                                                                       test agent::state::tests::test_commit_response_to_json ... ok                                                                               test agent::state::tests::test_get_response_to_json ... ok                                                                                  test agent::state::tests::serialize_round_trip_agent_state ... ok                                                                           test action::tests::action_wrapper_id ... ok                                                                                                test agent::state::tests::test_get_links_response_to_json ... ok                                                                            test action::tests::new_action_wrapper ... ok                                                                                               test agent::state::tests::test_link_entries_response_to_json ... ok                                                                         test agent::chain_store::tests::glob_query_test ... ok                                                                                      test context::tests::default_buffer_size_test ... ok                                                                                        test agent::state::tests::agent_state_new ... ok                                                                                            test agent::state::tests::agent_state_actions ... ok                                                                                        test context::tests::state_test ... ok                                                                                                      test agent::chain_store::tests::iterator_test ... ok                                                                                        test dht::dht_store::tests::get_headers_roundtrip ... ok                                                                                    test dht::dht_reducers::tests::does_not_add_link_for_missing_base ... ok                                                                    test agent::chain_store::tests::type_iterator_test ... ok                                                                                   test agent::state::tests::test_create_new_chain_header ... ok                                                                               test instance::tests::can_dispatch_and_wait ... ok                                                                                          test dht::dht_reducers::tests::can_add_links ... ok                                                                                         test agent::state::tests::test_reduce_commit_entry ... ok                                                                                   test instance::tests::default_buffer_size_test ... ok                                                                                       test agent::chain_store::tests::query_test ... ok                                                                                           test dht::dht_reducers::tests::reduce_hold_entry_test ... ok                                                                                test instance::tests::can_commit_agent ... ok                                                                                               test dht::dht_reducers::tests::can_remove_links ... FAILED                                                                                  test logger::tests::test_channel_logger ... ok                                                                                              test instance::tests::can_commit_dna ... ok                                                                                                 test logger::tests::test_test_logger ... ok                                                                                                 test logger::tests::simple_logger_smoke_test ... ok                                                                                         test instance::tests::can_process_action ... ok                                                                                             test network::reducers::get_entry::tests::reduce_get_entry_without_network_initialized ... ok                                               test network::reducers::get_links::tests::reduce_get_links_without_network_initialized ... ok                                               test network::tests::get_entry_roundtrip ... ignored                                                                                        test network::reducers::publish::tests::reduce_publish_test ... ok                                                                          test network::tests::get_links_roundtrip ... ignored                                                                                        test network::reducers::send_direct_message::tests::reduce_send_direct_message_timeout_test ... ok                                          test dht::dht_reducers::tests::reduce_hold_test ... ok                                                                                      test link_tests::tests::can_commit_link ... ok                                                                                              test link_tests::tests::can_commit_multilink ... ok                                                                                         test instance::tests::test_genesis_err ... ok                                                                                               test instance::tests::test_genesis_ok ... ok                                                                                                test instance::tests::test_missing_genesis ... ok                                                                                           test nucleus::actions::call_zome_function::tests::test_agent_as_token ... ok                                                                test nucleus::actions::call_zome_function::tests::test_call_signatures ... ok                                                               test nucleus::actions::call_zome_function::tests::test_make_cap_request_for_call ... ok                                                     test nucleus::actions::call_zome_function::tests::test_verify_grant ... ok                                                                  thread '' panicked at 'called `Result::unwrap()` on an `Err` value: "is_my_dna() called without DNA"', src\libcore\result.rs:997:5 test nucleus::actions::get_entry::tests::test_get_entry_from_dht_cas ... ok                                                                 test network::tests::get_entry_results_roundtrip ... ok                                                                                     test nucleus::reducers::init_application::tests::can_reduce_initialize_action ... ok                                                        test nucleus::reducers::return_initialization_result::tests::can_reduce_return_init_result_action ... ok                                    test nucleus::reducers::return_zome_function_result::tests::test_reduce_return_zome_function_result ... ok                                  test network::tests::get_entry_when_alone ... ok                                                                                            test network::tests::get_non_existant_entry ... FAILED                                                                                      test nucleus::ribosome::api::call::tests::test_call_no_zome ... ok                                                                          test nucleus::ribosome::api::call::tests::test_call_assigned ... ok                                                                         test nucleus::ribosome::api::call::tests::test_call_public ... ok                                                                           test network::tests::get_validation_package_roundtrip ... ok                                                                                test nucleus::ribosome::api::call::tests::test_validate_call_by_agent ... ok                                                                test nucleus::ribosome::api::call::tests::test_check_capability_transferable ... ok                                                         test nucleus::ribosome::api::call::tests::test_validate_call_public ... ok                                                                  test nucleus::ribosome::api::commit::tests::test_commit_round_trip ... ok                                                                   test nucleus::ribosome::api::get_entry::tests::test_get_round_trip ... ok                                                                   test nucleus::ribosome::api::init_globals::tests::test_init_globals ... ok                                                                  test dht::actions::add_link::tests::errors_when_link_base_not_present ... ok                                                                test dht::actions::add_link::tests::can_add_valid_link ... ok                                                                               test nucleus::ribosome::api::tests::defn_test ... ok                                                                                        test nucleus::ribosome::api::tests::test_from_str ... ok                                                                                    test nucleus::ribosome::api::get_links::tests::returns_list_of_links ... ok                                                                 test nucleus::ribosome::api::link_entries::tests::errors_with_wrong_tag ... ok                                                              test nucleus::actions::tests::can_instantiate_test_instance ... ok                                                                          test nucleus::ribosome::api::link_entries::tests::returns_ok_if_base_is_present ... ok                                                      test nucleus::ribosome::callback::genesis::tests::fail ... ok                                                                               test nucleus::ribosome::callback::tests::defn_test ... ok                                                                                   test nucleus::ribosome::callback::tests::test_from_str ... ok                                                                               test nucleus::ribosome::callback::genesis::tests::pass ... ok                                                                               test nucleus::ribosome::capabilities::tests::test_capability_request_new ... ok                                                             test nucleus::tests::call_ribosome_wrong_dna ... ok                                                                                         test nucleus::ribosome::callback::genesis::tests::not_implemented ... ok                                                                    test nucleus::tests::can_instantiate_nucleus_state ... ok                                                                                   test nucleus::ribosome::callback::receive::tests::implemented_with_null ... ok                                                              test nucleus::ribosome::api::link_entries::tests::works_with_linked_from_defined_link ... ok                                                test nucleus::tests::test_call_result_result ... ok                                                                                         test nucleus::ribosome::callback::receive::tests::not_implemented ... ok                                                                    test nucleus::tests::test_zome_call_eq ... ok                                                                                               test nucleus::tests::test_zome_call_result ... ok                                                                                           test nucleus::tests::test_zomefncall_same_as ... ok                                                                                         test persister::tests::persistence_round_trip ... ok                                                                                        test nucleus::tests::call_ribosome_wrong_function ... ok                                                                                    test nucleus::tests::call_wrong_zome_function ... ok                                                                                        test nucleus::actions::call_zome_function::tests::test_get_grant ... ok                                                                     test nucleus::tests::test_call_zome_function ... ok                                                                                         test nucleus::actions::build_validation_package::tests::test_building_validation_package_chain_headers ... ok                               test nucleus::actions::build_validation_package::tests::test_building_validation_package_chain_entries ... ok                               test nucleus::actions::build_validation_package::tests::test_building_validation_package_entry ... ok                                       test nucleus::actions::build_validation_package::tests::test_building_validation_package_chain_full ... ok                                  test nucleus::ribosome::api::call::tests::test_call_transferable ... test nucleus::ribosome::api::call::tests::test_call_transferable has been running for over 60 seconds
```

---

But it appears to be somewhat flaky. Running the same test suite several times also resulted in 

---

```

Finished release [optimized] target(s) in 0.45s                                                                                              Running target\release\deps\holochain_core-f48f6abc05c74485.exe                                                                                                                                                                                                                    running 99 tests                                                                                                                            test action::tests::action_wrapper_action ... ok                                                                                            test action::tests::action_wrapper_id ... ok                                                                                                test action::tests::action_wrapper_hash ... ok                                                                                              test action::tests::new_action_wrapper ... ok                                                                                               test action::tests::new_action ... ok                                                                                                       test agent::state::tests::test_commit_response_to_json ... ok                                                                               test agent::state::tests::serialize_round_trip_agent_state ... ok                                                                           test agent::state::tests::test_get_links_response_to_json ... ok                                                                            test agent::state::tests::test_link_entries_response_to_json ... ok                                                                         test agent::state::tests::test_get_response_to_json ... ok                                                                                  test context::tests::default_buffer_size_test ... ok                                                                                        test agent::chain_store::tests::glob_query_test ... ok                                                                                      test agent::state::tests::agent_state_new ... ok                                                                                            test agent::state::tests::agent_state_actions ... ok                                                                                        test agent::chain_store::tests::iterator_test ... ok                                                                                        test context::tests::state_test ... ok                                                                                                      test dht::dht_reducers::tests::does_not_add_link_for_missing_base ... ok                                                                    test agent::chain_store::tests::type_iterator_test ... ok                                                                                   test dht::dht_store::tests::get_headers_roundtrip ... ok                                                                                    test agent::state::tests::test_create_new_chain_header ... ok                                                                               test instance::tests::can_dispatch_and_wait ... ok                                                                                          test instance::tests::default_buffer_size_test ... ok                                                                                       test agent::state::tests::test_reduce_commit_entry ... ok                                                                                   test dht::dht_reducers::tests::can_add_links ... ok                                                                                         test agent::chain_store::tests::query_test ... ok                                                                                           test dht::dht_reducers::tests::reduce_hold_test ... ok                                                                                      test instance::tests::can_process_action ... ok                                                                                             test instance::tests::can_commit_dna ... ok                                                                                                 test logger::tests::test_channel_logger ... ok                                                                                              test logger::tests::test_test_logger ... ok                                                                                                 test dht::dht_reducers::tests::reduce_hold_entry_test ... ok                                                                                test logger::tests::simple_logger_smoke_test ... ok                                                                                         test dht::dht_reducers::tests::can_remove_links ... ok                                                                                      test instance::tests::can_commit_agent ... ok                                                                                               test network::reducers::get_links::tests::reduce_get_links_without_network_initialized ... ok                                               test network::tests::get_entry_roundtrip ... ignored                                                                                        test network::reducers::get_entry::tests::reduce_get_entry_without_network_initialized ... ok                                               test network::tests::get_links_roundtrip ... ignored                                                                                        test network::reducers::publish::tests::reduce_publish_test ... ok                                                                          test network::reducers::send_direct_message::tests::reduce_send_direct_message_timeout_test ... ok                                          test link_tests::tests::can_commit_link ... ok                                                                                              test link_tests::tests::can_commit_multilink ... ok                                                                                         test instance::tests::test_genesis_err ... ok                                                                                               test instance::tests::test_genesis_ok ... ok                                                                                                test instance::tests::test_missing_genesis ... ok                                                                                           thread '' panicked at 'called `Result::unwrap()` on an `Err` value: "is_my_dna() called without DNA"', src\libcore\result.rs:997:5 note: Run with `RUST_BACKTRACE=1` environment variable to display a backtrace.                                                              test nucleus::actions::call_zome_function::tests::test_agent_as_token ... ok                                                                test nucleus::actions::call_zome_function::tests::test_call_signatures ... ok                                                               test nucleus::actions::call_zome_function::tests::test_make_cap_request_for_call ... ok                                                     test nucleus::actions::call_zome_function::tests::test_verify_grant ... ok                                                                  test nucleus::actions::get_entry::tests::test_get_entry_from_dht_cas ... ok                                                                 test network::tests::get_entry_results_roundtrip ... ok                                                                                     test nucleus::reducers::init_application::tests::can_reduce_initialize_action ... ok                                                        test nucleus::reducers::return_initialization_result::tests::can_reduce_return_init_result_action ... ok                                    test nucleus::reducers::return_zome_function_result::tests::test_reduce_return_zome_function_result ... ok                                  test network::tests::get_entry_when_alone ... ok                                                                                            test network::tests::get_non_existant_entry ... FAILED                                                                                      test nucleus::ribosome::api::call::tests::test_call_no_zome ... ok                                                                          test network::tests::get_validation_package_roundtrip ... ok                                                                                test nucleus::ribosome::api::call::tests::test_call_assigned ... ok                                                                         test nucleus::ribosome::api::call::tests::test_call_public ... ok                                                                           test nucleus::ribosome::api::call::tests::test_validate_call_by_agent ... ok                                                                test nucleus::ribosome::api::call::tests::test_validate_call_public ... ok                                                                  test nucleus::ribosome::api::call::tests::test_call_transferable ... ok                                                                     test nucleus::ribosome::api::call::tests::test_check_capability_transferable ... ok                                                         test dht::actions::add_link::tests::errors_when_link_base_not_present ... ok                                                                test nucleus::ribosome::api::init_globals::tests::test_init_globals ... ok                                                                  test nucleus::ribosome::api::commit::tests::test_commit_round_trip ... ok                                                                   test nucleus::ribosome::api::get_entry::tests::test_get_round_trip ... ok                                                                   test nucleus::ribosome::api::tests::defn_test ... ok                                                                                        test nucleus::ribosome::api::tests::test_from_str ... ok                                                                                    test nucleus::ribosome::callback::genesis::tests::fail ... ok                                                                               test dht::actions::add_link::tests::can_add_valid_link ... ok                                                                               test nucleus::ribosome::api::link_entries::tests::errors_with_wrong_tag ... ok                                                              test nucleus::ribosome::api::get_links::tests::returns_list_of_links ... ok                                                                 test nucleus::actions::tests::can_instantiate_test_instance ... ok                                                                          test nucleus::ribosome::api::link_entries::tests::returns_ok_if_base_is_present ... ok                                                      test nucleus::ribosome::callback::tests::defn_test ... ok                                                                                   test nucleus::ribosome::callback::tests::test_from_str ... ok                                                                               test nucleus::ribosome::capabilities::tests::test_capability_request_new ... ok                                                             test nucleus::ribosome::callback::genesis::tests::pass ... ok                                                                               test nucleus::tests::call_ribosome_wrong_dna ... ok                                                                                         test nucleus::tests::can_instantiate_nucleus_state ... ok                                                                                   test nucleus::tests::test_call_result_result ... ok                                                                                         test nucleus::actions::build_validation_package::tests::test_building_validation_package_chain_entries ... ok                               test nucleus::ribosome::callback::genesis::tests::not_implemented ... ok                                                                    test nucleus::ribosome::api::link_entries::tests::works_with_linked_from_defined_link ... ok                                                test nucleus::tests::test_zome_call_eq ... ok                                                                                               test nucleus::tests::test_zome_call_result ... ok                                                                                           test nucleus::tests::test_zomefncall_same_as ... ok                                                                                         test persister::tests::persistence_round_trip ... ok                                                                                        test nucleus::ribosome::callback::receive::tests::implemented_with_null ... ok                                                              test nucleus::tests::call_wrong_zome_function ... ok                                                                                        test nucleus::ribosome::callback::receive::tests::not_implemented ... ok                                                                    test nucleus::tests::test_call_zome_function ... ok                                                                                         test nucleus::tests::call_ribosome_wrong_function ... ok                                                                                    test nucleus::actions::build_validation_package::tests::test_building_validation_package_chain_full ... ok                                  test nucleus::actions::build_validation_package::tests::test_building_validation_package_entry ... ok                                       test nucleus::actions::build_validation_package::tests::test_building_validation_package_chain_headers ... ok                               test nucleus::actions::call_zome_function::tests::test_get_grant ... test nucleus::actions::call_zome_function::tests::test_get_grant has b een running for over 60 seconds
```


 and 


```

Finished release [optimized] target(s) in 0.48s                                                                                              Running target\release\deps\holochain_core-f48f6abc05c74485.exe                                                                                                                                                                                                                    running 99 tests                                                                                                                            test action::tests::action_wrapper_id ... ok                                                                                                test action::tests::action_wrapper_action ... ok                                                                                            test action::tests::new_action_wrapper ... ok                                                                                               test action::tests::action_wrapper_hash ... ok                                                                                              test action::tests::new_action ... ok                                                                                                       test agent::state::tests::test_commit_response_to_json ... ok                                                                               test agent::state::tests::serialize_round_trip_agent_state ... ok                                                                           test agent::state::tests::test_get_links_response_to_json ... ok                                                                            test agent::state::tests::test_get_response_to_json ... ok                                                                                  test agent::state::tests::test_link_entries_response_to_json ... ok                                                                         test context::tests::default_buffer_size_test ... ok                                                                                        test agent::chain_store::tests::glob_query_test ... ok                                                                                      test agent::state::tests::agent_state_new ... ok                                                                                            test agent::state::tests::agent_state_actions ... ok                                                                                        test context::tests::state_test ... ok                                                                                                      test dht::dht_reducers::tests::does_not_add_link_for_missing_base ... ok                                                                    test dht::dht_store::tests::get_headers_roundtrip ... ok                                                                                    test agent::state::tests::test_create_new_chain_header ... ok                                                                               test agent::chain_store::tests::iterator_test ... ok                                                                                        test agent::state::tests::test_reduce_commit_entry ... ok                                                                                   test instance::tests::can_dispatch_and_wait ... ok                                                                                          test instance::tests::default_buffer_size_test ... ok                                                                                       test agent::chain_store::tests::type_iterator_test ... ok                                                                                   test dht::dht_reducers::tests::can_add_links ... ok                                                                                         test instance::tests::can_commit_agent ... ok                                                                                               test dht::dht_reducers::tests::reduce_hold_test ... ok                                                                                      test dht::dht_reducers::tests::can_remove_links ... ok                                                                                      test dht::dht_reducers::tests::reduce_hold_entry_test ... ok                                                                                test logger::tests::test_channel_logger ... ok                                                                                              test logger::tests::test_test_logger ... ok                                                                                                 test logger::tests::simple_logger_smoke_test ... ok                                                                                         test instance::tests::can_process_action ... ok                                                                                             test instance::tests::can_commit_dna ... ok                                                                                                 test network::reducers::get_entry::tests::reduce_get_entry_without_network_initialized ... ok                                               test network::reducers::get_links::tests::reduce_get_links_without_network_initialized ... ok                                               test network::tests::get_entry_roundtrip ... ignored                                                                                        test network::reducers::publish::tests::reduce_publish_test ... ok                                                                          test network::tests::get_links_roundtrip ... ignored                                                                                        test network::reducers::send_direct_message::tests::reduce_send_direct_message_timeout_test ... ok                                          test agent::chain_store::tests::query_test ... ok                                                                                           test link_tests::tests::can_commit_link ... ok                                                                                              test link_tests::tests::can_commit_multilink ... ok                                                                                         test instance::tests::test_genesis_err ... ok                                                                                               test instance::tests::test_genesis_ok ... ok                                                                                                test instance::tests::test_missing_genesis ... ok                                                                                           test nucleus::actions::call_zome_function::tests::test_agent_as_token ... ok                                                                test nucleus::actions::call_zome_function::tests::test_call_signatures ... ok                                                               test nucleus::actions::call_zome_function::tests::test_make_cap_request_for_call ... ok                                                     test nucleus::actions::call_zome_function::tests::test_verify_grant ... ok                                                                  thread '' panicked at 'called `Result::unwrap()` on an `Err` value: "is_my_dna() called without DNA"', src\libcore\result.rs:997:5 note: Run with `RUST_BACKTRACE=1` environment variable to display a backtrace.                                                              test nucleus::actions::get_entry::tests::test_get_entry_from_dht_cas ... ok                                                                 test network::tests::get_entry_results_roundtrip ... ok                                                                                     test nucleus::reducers::init_application::tests::can_reduce_initialize_action ... ok                                                        test nucleus::reducers::return_initialization_result::tests::can_reduce_return_init_result_action ... ok                                    test nucleus::reducers::return_zome_function_result::tests::test_reduce_return_zome_function_result ... ok                                  test network::tests::get_entry_when_alone ... ok                                                                                            test network::tests::get_non_existant_entry ... FAILED                                                                                      test nucleus::ribosome::api::call::tests::test_call_no_zome ... ok                                                                          test nucleus::ribosome::api::call::tests::test_call_public ... ok                                                                           test network::tests::get_validation_package_roundtrip ... ok                                                                                test nucleus::ribosome::api::call::tests::test_validate_call_by_agent ... ok                                                                test nucleus::ribosome::api::call::tests::test_check_capability_transferable ... ok                                                         test nucleus::ribosome::api::call::tests::test_call_transferable ... ok                                                                     test nucleus::ribosome::api::call::tests::test_validate_call_public ... ok                                                                  test dht::actions::add_link::tests::errors_when_link_base_not_present ... ok                                                                test nucleus::ribosome::api::commit::tests::test_commit_round_trip ... ok                                                                   test nucleus::ribosome::api::get_entry::tests::test_get_round_trip ... ok                                                                   test nucleus::ribosome::api::init_globals::tests::test_init_globals ... ok                                                                  test dht::actions::add_link::tests::can_add_valid_link ... ok                                                                               test nucleus::ribosome::api::tests::defn_test ... ok                                                                                        test nucleus::ribosome::api::tests::test_from_str ... ok                                                                                    test nucleus::actions::tests::can_instantiate_test_instance ... ok                                                                          test nucleus::ribosome::api::link_entries::tests::errors_with_wrong_tag ... ok                                                              test nucleus::ribosome::api::get_links::tests::returns_list_of_links ... ok                                                                 test nucleus::ribosome::callback::genesis::tests::fail ... ok                                                                               test nucleus::ribosome::api::link_entries::tests::returns_ok_if_base_is_present ... ok                                                      test nucleus::ribosome::callback::tests::defn_test ... ok                                                                                   test nucleus::ribosome::callback::tests::test_from_str ... ok                                                                               test nucleus::ribosome::capabilities::tests::test_capability_request_new ... ok                                                             test nucleus::ribosome::callback::genesis::tests::not_implemented ... ok                                                                    test nucleus::tests::call_ribosome_wrong_dna ... ok                                                                                         test nucleus::ribosome::callback::receive::tests::implemented_with_null ... ok                                                              test nucleus::tests::can_instantiate_nucleus_state ... ok                                                                                   test nucleus::tests::test_call_result_result ... ok                                                                                         test nucleus::ribosome::api::link_entries::tests::works_with_linked_from_defined_link ... ok                                                test nucleus::tests::test_zome_call_eq ... ok                                                                                               test nucleus::tests::test_zome_call_result ... ok                                                                                           test nucleus::tests::test_zomefncall_same_as ... ok                                                                                         test nucleus::ribosome::callback::receive::tests::not_implemented ... ok                                                                    test nucleus::ribosome::callback::genesis::tests::pass ... ok                                                                               test persister::tests::persistence_round_trip ... ok                                                                                        test nucleus::tests::call_ribosome_wrong_function ... ok                                                                                    test nucleus::tests::test_call_zome_function ... ok                                                                                         test nucleus::tests::call_wrong_zome_function ... ok                                                                                        test nucleus::actions::build_validation_package::tests::test_building_validation_package_chain_entries ... ok                               test nucleus::actions::build_validation_package::tests::test_building_validation_package_chain_headers ... ok                               test nucleus::actions::build_validation_package::tests::test_building_validation_package_chain_full ... ok                                  test nucleus::actions::build_validation_package::tests::test_building_validation_package_entry ... ok                                       test nucleus::actions::call_zome_function::tests::test_get_grant ... test nucleus::actions::call_zome_function::tests::test_get_grant has b een running for over 60 seconds                                                                                                             test nucleus::ribosome::api::call::tests::test_call_assigned ... test nucleus::ribosome::api::call::tests::test_call_assigned has been runn ing for over 60 seconds
```

Hope this helps

lucksus commented:
@zippy, I have resolved conflicts with develop and [refactored get_public_token()](https://github.com/holochain/holochain-rust/pull/895/commits/dc31acf696686586c4782a4343835ec68cd2287b).

I do see a failing test though locally, even without my refactor (but after the develop merge). Let's see if CI picks that up as well.

I have:
```2019-02-26 13:47:37:bridge-caller: debug/net/handle: HandleStoreEntry: EntryData { dna_address: HashString("Qma2zPiLSuAS5U7tafQsgaGvGBUV1taEsUGsqzXjnuc8mc"), provider_agent_id: "qQg205btpxl7U6fwV6wUfMAz_xxY8PLOMmRqHPMDq31b45sSI83BT6aI-MPDvJDyXuAjQQdq_ns1V6fbe6HLXY5K5wPw", entry_address: HashString("qQg205btpxl7U6fwV6wUfMAz_xxY8PLOMmRqHPMDq31b45sSI83BT6aI-MPDvJDyXuAjQQdq_ns1V6fbe6HLXY5K5wPw"), entry_content: Object({"entry": Object({"AgentId": Object({"nick": String("Holo Tester 3"), "key": String("qQg205btpxl7U6fwV6wUfMAz_xxY8PLOMmRqHPMDq31b45sSI83BT6aI-MPDvJDyXuAjQQdq_ns1V6fbe6HLXY5K5wPw")})}), "header": Object({"entry_type": String("AgentId"), "entry_address": String("qQg205btpxl7U6fwV6wUfMAz_xxY8PLOMmRqHPMDq31b45sSI83BT6aI-MPDvJDyXuAjQQdq_ns1V6fbe6HLXY5K5wPw"), "provenances": Array([Array([String("qQg205btpxl7U6fwV6wUfMAz_xxY8PLOMmRqHPMDq31b45sSI83BT6aI-MPDvJDyXuAjQQdq_ns1V6fbe6HLXY5K5wPw"), String("89+Wfu0SIar2Kyziao/mJhYh7sXqpq4EL7zS65SDX060b0BiIJqAFZ9mygpjlliUs1vlgNwYl8Wgfasx63A6DA==")])]), "link": String("QmRJW4cZiK3YqUUwZfXaVH7RDVWEDQ26HJpQUjgHB5Gthe"), "link_same_type": Null, "link_crud": Null, "timestamp": String("2019-02-26T12:47:35+00:00")})}) }
2019-02-26 13:47:37:bridge-caller: err/zome: invoke_call failed to deserialize: JsonString("{\"instance_handle\":\"test-callee\",\"zome_name\":\"greeter\",\"cap\":{\"cap_token\":\"token\",\"caller\":null,\"signature\":{}},\"fn_name\":\"hello\",\"fn_args\":\"{}\"}")
2019-02-26 13:47:37:bridge-caller: debug/dna: '"called `Result::unwrap()` on an `Err` value: Internal(\"Argument deserialization failed\")"'
2019-02-26 13:47:37:bridge-caller: debug/dna: '"panic occurred in file 'src/libcore/result.rs' at line 997"'
```

in `conductor::base::tests::basic_bridge_call_roundtrip`.

lucksus commented on core/src/nucleus/actions/call_zome_function.rs:
```diff
@@ -141,32 +117,162 @@ pub async fn call_zome_function(
-    if is_token_the_agent(context.clone(), &fn_call.cap) {
-        return true;
+fn is_token_the_agent(context: Arc<Context>, request: &CapabilityRequest) -> bool {
+    context.agent_id.key == request.cap_token.to_string()
```
Should this be:
```suggestion
    context.agent_id.address()? == request.cap_token.to_string()
```
?

lucksus commented on core/src/nucleus/actions/call_zome_function.rs:
```diff
@@ -201,19 +307,252 @@ impl Future for CallResultFuture {
+
+        let mut cap_functions = CapFunctions::new();
+        cap_functions.insert("test_zome".to_string(), vec![String::from("test")]);
+        let grant =
```
Should we split the below into a different test function since it's a different case?

lucksus commented on core/src/nucleus/ribosome/callback/validate_entry.rs:
```diff
@@ -113,9 +113,14 @@ fn validate_link_entry(
         validation_data,
     };
+
     let call = ZomeFnCall::new(
```
Hm, I really think this should be a `WasmCallData::DirectCall`, no?

lucksus replied:
Ah, ZomeFnCall doesn't know about `WasmCallData`.. hm, I'm a bit confused now..

lucksus commented on core/src/nucleus/ribosome/callback/validate_entry.rs:
```diff
@@ -113,9 +113,14 @@ fn validate_link_entry(
     let call = ZomeFnCall::new(
         &link_definition_path.zome_name,
-        None,
+        make_internal_capability_request(
```
Could this stay `None` if `run_validation_callback` below would do a `DirectCall`, which it should?

zippy replied:
Hmm.  Maybe this makes sense.  I'll look at that.

lucksus commented on core/src/nucleus/ribosome/callback/validation_package.rs:
```diff
@@ -39,9 +39,13 @@ pub fn get_validation_package_definition(
             let call = ZomeFnCall::new(
                 &zome_name,
-                None,
+                make_internal_capability_request(
```
Same question as above. These `__hdk_*` calls should not need a capability token, imho (-> `DirectCall`)

lucksus commented on core/src/nucleus/ribosome/capabilities.rs:
```diff
@@ -0,0 +1,46 @@
+pub struct CapabilityRequest {
+    pub cap_token: Address,
+    pub provenance: Provenance,
+}
```
Inquiry: Should `CapabilityRequest` include the whole function call as well?

Context: We'll need to define a serialized representation of a signed call that we can create and sign in UIs and other clients and also deserialize and check in core. Should this become that?

zippy replied:
I thought about that.  In the general case the provenance should point to what it's a provenance for, and the request could be for things other than zome calls.  Because right now there is only one thing you can request you always know contextually what is signed in the provenance, so it's not needed.

lucksus commented on doc/architecture/decisions/0017-capabilities.md:
```diff
@@ -0,0 +1,94 @@
+# Capabilities & Security ADR
+
+Date: 2019-02-xx
```
Should be set before merging.
```suggestion
Date: 2019-02-26
```

lucksus commented on doc/architecture/decisions/0017-capabilities.md:
```diff
@@ -0,0 +1,94 @@
+
+1. Agent bundles function name, parameters and a timestamp into a call request block, and signs it with the agent's private key, and sends it to Conductor along with any other parameters necessary for routing to the correct instance over what ever interface is being used.
+2. Conductor creates a CapabilityRequest structure and passes it into holochain Core.
+3. Core loads CapabilityGrant from chain by it's address, and checks validity according to the grant's parameters, returing a CapabilityCheckFailed error, or calling the zome function if successfull.  It may also check the timestamp of the call to make-sure it's within a reasonble window to prevent some re-play attacks.  Additionally, if complete security from replay attacks is necessary we may i ...
```
```suggestion
3. Core loads CapabilityGrant from chain by it's address, and checks validity according to the grant's parameters, returning a CapabilityCheckFailed error, or calling the zome function if successfull.  It may also check the timestamp of the call to make-sure it's within a reasonble window to prevent some re-play attacks.  Additionally, if complete security from replay attacks is necessary we may implement an additional handshake where the agent makes a "pre-call" indicating the desire to make a zome call.  In that case the Conductor would have to pass this request into core where a nonce could be generated that the client has to include in the call request block. Note that this also requires implementing an ephemeral store in core, something that's on our development path.
```

lucksus commented on doc/holochain_101/src/zome/zome_functions.md:
````diff
@@ -139,6 +139,24 @@ define_zome! {
+}
+```
+
+In this example, `hc_public` is the reserved trait name which creates a `Public` Capbility-type grant at genesis time for access to the `read_post` function.  Additionally it names an `authoring` trait the `create_post` and `update_post` functions.
````
```suggestion
In this example, `hc_public` is the reserved trait name which creates a `Public` Capbility-type grant at genesis time for access to the `read_post` function.  Additionally it names an `authoring` trait for the `create_post` and `update_post` functions.
```

willemolding commented:
Ok @zippy with the latest changes it still hangs on my windows machine with
```
zome_function::tests::test_get_grant has been running for over 60 seconds
```
which is strange since it seems to be ok in CI.

willemolding commented:
aaand there we go 👍 
```
test result: ok. 98 passed; 0 failed; 2 ignored; 0 measured; 0 filtered out
```

willemolding commented:
Ok lucky you got me to run it again. It still is flaky! From 10 tests runs it gave this error 4 times
```

failures:                                                                                                                                                                                                     ---- nucleus::ribosome::api::get_links::tests::returns_list_of_links stdout ----                       thread 'nucleus::ribosome::api::get_links::tests::returns_list_of_links' panicked at '                  call_result = 'JsonString("{\"ok\":true,\"value\":\"{\\\"addresses\\\":[]}\",\"error\":\"null\"}")'      ordering1 = 'JsonString("{\"ok\":true,\"value\":\"{\\\"addresses\\\":[\\\"QmQjeg9grsaGkdre5RWvHRf7v JYxid5wcF4ZLSUQ9vQy5N\\\",\\\"QmdGW5XTdvLvTnotVYiQaM5qXviSbpaeYzsCMyTek4MiVh\\\"]}\",\"error\":\"null\ "}")'                                                                                                     ordering2 = 'JsonString("{\"ok\":true,\"value\":\"{\\\"addresses\\\":[\\\"QmdGW5XTdvLvTnotVYiQaM5qX viSbpaeYzsCMyTek4MiVh\\\",\\\"QmQjeg9grsaGkdre5RWvHRf7vJYxid5wcF4ZLSUQ9vQy5N\\\"]}\",\"error\":\"null\ "}")'', core\src\nucleus\ribosome\api\get_links.rs:141:9
```

willemolding commented:
still intermittently failing (approx 50% of the time) with the same error above
